### PR TITLE
Role focus adjustments

### PR DESCRIPTION
**Pull request by JFranklin3000**
##### In this PR
- Added Foci for GI Joe and Transformers Roles
- Added all Roles and Focus for GI Joe CRB

##### Testing
-fvtt package pack -n "gijcrbitems" 
Validate dropping Focus on to actors with a GI Joe Role.

### CODE REVIEW

File: module/helpers/utils.mjs [hunk 1]
```diff
@@ -165,7 +165,7 @@ export async function createItemCopies(items, owner, type, parentItem, lastProce
   for (const [key, item] of Object.entries(items)) {
     if (item.type == type) {
       const createNewItem =
-        parentItem.type != "role"
+        (parentItem.type != "role" && parentItem.type != "focus")
```
Comment (phildominguez) [inline module/helpers/utils.mjs:168]:
```suggestion
        !["role", "focus"].includes(parentItem.type)
```
Not tested

File: module/helpers/utils.mjs [hunk 2]
```diff
@@ -446,6 +448,7 @@ export function deleteAttachmentsForItem(item, actor, previousLevel=null) {
  * @returns {Number} totalChange The number of level changes.
  */
 export async function roleValueChange(currentLevel, arrayLevels, lastProcessedLevel=null) {
+
```
Comment (phildominguez) [inline module/helpers/utils.mjs:451]:
```suggestion
```

File: module/sheet-handlers/role-handler.mjs [hunk 3]
```diff
@@ -76,6 +178,11 @@ export class RoleHandler {
       });
     }
 
+    if(focus[0]) {
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:181]:
```suggestion
    if (focus[0]) {
```

File: module/sheets/actor-sheet.mjs [hunk 4]
```diff
@@ -158,6 +158,7 @@ export class Essence20ActorSheet extends ActorSheet {
     const bonds = [];
     const contacts = [];
     const features = []; // Used by Zords
+    const foci = [];
```
Comment (phildominguez) [inline module/sheets/actor-sheet.mjs:161]: I wasn't sure so I [looked it up](https://www.dictionary.com/browse/focus), but "focuses" is perfectly valid. I think we should go with that since it's more clear.

File: module/sheets/item-sheet.mjs [hunk 5]
```diff
@@ -93,6 +93,9 @@ export class Essence20ItemSheet extends ItemSheet {
     // Delete Role Perk from Influence
     html.find('.rolePerk-delete').click(this._onObjectDelete.bind(this, ".perk"));
 
+    // Delete Role from Focus
+    html.find('.role-delete').click(this._onObjectDelete.bind(this, ".role"));
+
```
Comment (phildominguez) [inline module/sheets/item-sheet.mjs:98]:
```suggestion
```

File: module/sheet-handlers/role-handler.mjs [hunk 6]
```diff
@@ -16,6 +18,105 @@ export class RoleHandler {
     this._actor = actorSheet.actor;
   }
 
+  async focusUpdate(focus, dropFunc){
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:21]:
```suggestion
  async focusUpdate(focus, dropFunc) {
```

File: module/sheet-handlers/role-handler.mjs [hunk 7]
```diff
@@ -16,6 +18,105 @@ export class RoleHandler {
...
+    }
+
+    // Characters can only have one Focus
+    if (hasFocus) {
+      ui.notifications.error(game.i18n.format(game.i18n.localize('E20.FocusMultipleError')));
+      return false;
+    }
+
+    if (role[0]) {
+      if (role[0].flags.core.sourceId == attachedRole[0].uuid) {
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:43]: I would do the failure versions of these checks individually to help reduce complexity here. I can elaborate on discord if needed.

File: module/sheet-handlers/role-handler.mjs [hunk 8]
```diff
@@ -16,6 +18,105 @@ export class RoleHandler {
...
+          await this._showEssenceDialog(focus, dropFunc);
+        } else {
+          const newFocusList = await dropFunc();
+          const newFocus = newFocusList[0];
+          await this._actor.update({
+            "system.focusEssence": newFocus.system.essences[0],
+          });
+          await setFocusValues(newFocus, this._actor);
+        }
+
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:54]:
```suggestion
```

File: module/sheet-handlers/role-handler.mjs [hunk 9]
```diff
@@ -16,6 +18,105 @@ export class RoleHandler {
...
+        ui.notifications.error(game.i18n.format(game.i18n.localize('E20.FocusRoleMismatchError')));
+        return false;
+      }
+    } else {
+      ui.notifications.error(game.i18n.format(game.i18n.localize('E20.FocusNoRoleError')));
+      return false;
+    }
+  }
+
+  async _showEssenceDialog(focus, dropFunc) {
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:65]: I feel like we have a bunch of these now, maybe we can consolidate? Can be a different PR. I think this should be named more specifically though if it's only for focuses.

File: lang/en.json [hunk 10]
```diff
@@ -210,6 +210,8 @@
     "EnergonRed": "Red",
     "EnergonSynthEn": "Synth-En",
     "EssenceAny": "Any",
+    "EssenceBoost": "Focus Essence",
```
Comment (phildominguez) [inline lang/en.json:213]: I'm not seeing this being used

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: lang/en.json [hunk 11]
```diff
@@ -229,6 +231,10 @@
     "FactionDecepticons": "Decepticons",
     "FactionOther": "Other",
     "Focus": "Focus",
+    "FocusDrop": "Drop a Focus here",
+    "FocusMultipleError": "You have already selected a Focus.",
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply lang/en.json:235]: changed to consistent verbiage.

File: module/helpers/utils.mjs (hunk 1 above)
Comment (JFranklin3000) [reply module/helpers/utils.mjs:168]: Now it is tested and working.

File: module/sheet-handlers/role-handler.mjs (hunk 3 above)
Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs:181]: fixed

File: module/sheets/actor-sheet.mjs (hunk 4 above)
Comment (JFranklin3000) [reply module/sheets/actor-sheet.mjs:161]: Sounds good to me I went back and forth on this one anyway. All changed.

File: module/sheets/item-sheet.mjs (hunk 5 above)
Comment (JFranklin3000) [reply module/sheets/item-sheet.mjs:98]: removed

File: module/sheet-handlers/role-handler.mjs (hunk 6 above)
Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs:21]: fixed

File: module/sheet-handlers/role-handler.mjs (hunk 8 above)
Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs:54]: fixed

File: lang/en.json (hunk 10 above)
Comment (JFranklin3000) [reply lang/en.json:213]: Must have been removed at some point, also removed.

File: module/sheet-handlers/role-handler.mjs (hunk 9 above)
Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs:65]: This is the same function we use for Origins at least, maybe Alterations as well.

Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs:65]: Nvm, that was the same dialog with a different function. we can definitely take a look.

File: module/sheet-handlers/role-handler.mjs (hunk 7 above)
Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs:43]: Changed

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/helpers/utils.mjs [hunk 12]
```diff
@@ -521,6 +521,13 @@ export async function setRoleValues(role, actor, newLevel=null, previousLevel=nu
   }
 }
 
+/**
+ * Handles setting the values and items for an actor's focus
+ * @param {Object} focus The actor's Focus
```
Comment (phildominguez) [inline module/helpers/utils.mjs:526]:
```suggestion
 * @param {Focus} focus The actor's Focus
```

File: module/sheet-handlers/role-handler.mjs [hunk 13]
```diff
@@ -79,14 +90,19 @@ export class RoleHandler {
...
     ).render(true);
   }
 
-  async _showFocusSkillDialog(options, dropFunc) {
+  /**
+   *Handles writing the selected Essence to the actor.
+  * @param {Object} options    The options resulting from _showFocusSkillDialog()
+   * @param {Function} dropFunc The drop Function that will be used to complete the drop of the focus
+   */
+  async _focusStatUpdate(options, dropFunc) {
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:105]:
```suggestion
  /**
   *Handles writing the selected Essence to the actor.
   * @param {Object} options    The options resulting from _showFocusSkillDialog()
   * @param {Function} dropFunc The drop Function that will be used to complete the drop of the focus
   */
  async _focusStatUpdate(options, dropFunc) {
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

**Review by phildominguez — APPROVED** [review]
